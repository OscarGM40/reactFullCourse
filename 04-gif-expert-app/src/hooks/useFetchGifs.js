import { useEffect, useState } from "react";
import { getGifs } from "../helpers/getGifs";


//este hook almacena un objeto como estado.El objeto tiene dos props,un array llamado data y un boleano para el loading.El array lo llenara getGifs en base al valor del input(category).
export const useFetchGifs = (category) => {
  
   const [state, setState] = useState({
    data: [],
    loading: true,
  });


  useEffect(() => {
        
      setTimeout(() => {
           
           getGifs(category)
           .then(imgs => //imgs es el array que devuelve getGifs.
              setState({
                 data: imgs,
                 loading: false
              })
           );
        },200)

    }, [category]); 


  return state; //state es un objeto con un array y un boleano <- es el estado de este custom hook.El setter sólo lo usé aqui
};
