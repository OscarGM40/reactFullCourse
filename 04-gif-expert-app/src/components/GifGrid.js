import React from "react";
import { useFetchGifs } from "../hooks/useFetchGifs";
import GifGridItem from "./GifGridItem";
import PropTypes from 'prop-types';

const GifGrid = ({ category }) => {

  const { data:images, loading } = useFetchGifs(category);

  return (
    <>
      <h3>{category}</h3>

      {loading && <p className="animate__animated animate__flash">Loading...</p>}

      <div className="card-grid">
        {images.map((img) => (
          // console.log(img) con el operador spread mando todas
          // despues recojo solo las que quiera.Muy Ãºtil
          <GifGridItem key={img.id} {...img} />
        ))}
      </div>
    </>
  );
};
GifGrid.propTypes = {
  category: PropTypes.string.isRequired,
}


export default GifGrid;
