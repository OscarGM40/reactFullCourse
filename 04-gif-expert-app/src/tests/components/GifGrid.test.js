import React from 'react';
import GifGrid from '../../components/GifGrid';
import { shallow } from 'enzyme';
import { useFetchGifs } from '../../hooks/useFetchGifs';

jest.mock('../../hooks/useFetchGifs');

describe('Pruebas en componente GifGrid', () => {
   
   const category = "cat";
      
   test('Debe de hacer match con el snapshot', () => {
      
      useFetchGifs.mockReturnValue({ 
         data: [], 
         loading: true,
      })
      
      const wrapper = shallow(<GifGrid category={category} />);
      expect(wrapper).toMatchSnapshot();
   })
   
   test('debe de mostrar items cuando se cargan imagenes con useFecthGifs', () => {
      //un mock es un se√±uelo
      
      const gifs = [{ 
         id: 'ABC',
         url: 'http://localhost/cualquiercosa/jpg', 
         title: 'Cualquier cosa'
      }]
      
      useFetchGifs.mockReturnValue({ 
         data: gifs, 
         loading: false,
      })
      
         const wrapper = shallow(<GifGrid category={category} />);
         expect(wrapper).toMatchSnapshot();
         expect( wrapper.find('p').exists() ).toBeFalsy();

         expect( wrapper.find('GifGridItem').length ).toBe( gifs.length ) 
   })
   
})
