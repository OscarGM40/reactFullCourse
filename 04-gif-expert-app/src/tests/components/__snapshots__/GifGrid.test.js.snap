// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en componente GifGrid Debe de hacer match con el snapshot 1`] = `
<Fragment>
  <h3>
    cat
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`Pruebas en componente GifGrid debe de mostrar items cuando se cargan imagenes con useFecthGifs 1`] = `
<Fragment>
  <h3>
    cat
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Cualquier cosa"
      url="http://localhost/cualquiercosa/jpg"
    />
  </div>
</Fragment>
`;
